File: leetcode-75/09-binary-tree-dfs/37-longest-zigzag-path-in-a-binary-tree/solution.py
"""
LeetCode 1372. Longest ZigZag Path in a Binary Tree
https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/

Approach: post-order DFS returning, for each node, the length of the
longest zigzag path ending there via a move to the left and via a
move to the right. Since a zigzag must alternate direction,
left(node) = 1 + right(node.left) and right(node) = 1 + left(node.right).
Track the global max at every node.

Time:  O(n) - every node is visited once
Space: O(h) - recursion stack, h = tree height
"""

from typing import Optional, Tuple


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        self.max_len = 0

        def dfs(node: Optional[TreeNode]) -> Tuple[int, int]:
            if node is None:
                return (-1, -1)

            _, left_from_right = dfs(node.left)
            right_from_left, _ = dfs(node.right)

            left = 1 + left_from_right
            right = 1 + right_from_left

            self.max_len = max(self.max_len, left, right)

            return (left, right)

        dfs(root)

        return self.max_len


if __name__ == "__main__":
    sol = Solution()
    tree = TreeNode(
        1,
        TreeNode(1, None, TreeNode(1, TreeNode(1, None, TreeNode(1)), TreeNode(1))),
        TreeNode(1),
    )
    print(sol.longestZigZag(tree))  # 4

File: leetcode-75/09-binary-tree-dfs/37-longest-zigzag-path-in-a-binary-tree/solution.ts
/**
 * LeetCode 1372. Longest ZigZag Path in a Binary Tree
 * https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/
 *
 * Approach: post-order DFS returning, for each node, the length of the
 * longest zigzag path ending there via a move to the left and via a
 * move to the right. Since a zigzag must alternate direction,
 * left(node) = 1 + right(node.left) and right(node) = 1 + left(node.right).
 * Track the global max at every node.
 *
 * Time:  O(n) - every node is visited once
 * Space: O(h) - recursion stack, h = tree height
 */
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function longestZigZag(root: TreeNode | null): number {
  let maxLen = 0;

  // Returns [longest ending with a move to the left, longest ending with a move to the right]
  function dfs(node: TreeNode | null): [number, number] {
    if (node === null) return [-1, -1];

    const [, leftFromRight] = dfs(node.left);
    const [rightFromLeft] = dfs(node.right);

    const left = 1 + leftFromRight;
    const right = 1 + rightFromLeft;

    maxLen = Math.max(maxLen, left, right);

    return [left, right];
  }

  dfs(root);

  return maxLen;
}

// Example usage:
// const tree = new TreeNode(1, new TreeNode(1, null, new TreeNode(1, new TreeNode(1, null, new TreeNode(1)), new TreeNode(1))), new TreeNode(1));
// console.log(longestZigZag(tree)); // 4

export { longestZigZag, TreeNode };
